### PR TITLE
TRUNK-5999: Add more filtering options to Diagnosis service

### PR DESCRIPTION
**Pull request by hgrsd**
## Description of what I changed
Creates two new methods in the `DiagnosisService` and corresponding DAOs:
- `getDiagnosesByEncounter`
- `getDiagnosesByVisit`

Each of these methods can filter by diagnoses that are primary or confirmed.

This backports the behaviour for `Visit`s in https://github.com/icrc-psousa/openmrs-module-emrapi/blob/RA-1714/api-2.2/src/main/java/org/openmrs/module/emrapi/diagnosis/EmrDiagnosisDAOImpl2_2.java into `openmrs-core` and introduces the same behaviour with regard to `Encounter`s.

This PR deprecates `DiagnosisService#getPrimaryDiagnoses` as the behaviour is now contained in `DiagnosisService#getDiagnosesByEncounter` which allows filtering for primary diagnoses; the same goes for `DiagnosisDAO#getDiagnoses`. I did not remove the methods for backwards compatibility reasons.

## Issue I worked on
https://issues.openmrs.org/browse/TRUNK-5999

## Checklist: I completed these to help reviewers :)
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

- [X] I ran `mvn clean package` right before creating this pull request and

- [X] All new and existing **tests passed**.

- [X] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by hgrsd** [comment]
@sherrif10 Hello :wave: As discussed on [the ticket](https://issues.openmrs.org/browse/TRUNK-5999) I've opened the PR to implement the behaviour described. Let me know what the next steps are. Thank you!

File: api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java [hunk 1]
```diff
@@ -58,20 +59,26 @@
...
+	 * primary or confirmed diagnoses.
 	 *
-	 * @param encounter the specific encounter to get the diagnoses for.
-	 * @return list of diagnoses for an encounter
+	 * @param encounter the encounter for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
+	 * @param confirmedOnly whether to return only confirmed diagnoses
+	 * @return the list of (primary, confirmed) diagnoses for the given encounter
 	 */
-	List<Diagnosis> getDiagnoses(Encounter encounter);
```
Comment (HerbertYiga) [inline api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:66]: thanks @hgrsd   are you considering things like backward compatibility? Don't you think that removing this will break the modules that depend on it?

Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:66]: Hi @HerbertYiga! Thanks for your comment. From what I understood from the developer guidelines, DAO methods can be removed without marking them as deprecated because no external modules must depend on them. The method wasn't used by any code in `openmrs-core`. Is that right?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:66]: They shouldn't depend on DAO classes but we have zero guarantees that they aren't :)

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 2]
```diff
@@ -14,22 +14,21 @@
 
 import org.hibernate.Query;
 import org.hibernate.SessionFactory;
-import org.openmrs.Diagnosis;
-import org.openmrs.Encounter;
-import org.openmrs.Patient;
+import org.openmrs.*;
```
Comment (BartlomiejRasztabiga) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:17]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

File: api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java (hunk 1 above)
Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:66]: Thanks. Clearly I was too enthusiastic getting rid of it - I'll bring it back for backwards compat.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 2 above)
Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:17]: Thanks - I think my IDE was not configured properly. Apologies.

File: api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java (hunk 1 above)
Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:66]: FYI, here is the docs I used which pointed to DAO methods being OK to remove: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Deprecation. I wonder if it's worth updating it to discourage doing so?

**Comment by hgrsd** [comment]
@BartlomiejRasztabiga @HerbertYiga would you be able to provide another review after the changes I made? thanks in advance!

**Review by BartlomiejRasztabiga — APPROVED** [review]

**Comment by hgrsd** [comment]
Also: is there anyone specifically I need to tag on this PR to get the CI pipeline to run?

**Comment by hgrsd** [comment]
@sherrif10 are there any specific people that I should tag on this PR for review and to run CI?

**Comment by hgrsd** [comment]
@dkayiwa @ibacher would one of you be able to provide a review and allow a CI run on my branch?

**Comment by dkayiwa** [comment]
@hgrsd i have just allowed CI to run your pull request.

**Comment by hgrsd** [comment]
> @hgrsd i have just allowed CI to run your pull request.

Thanks @dkayiwa, much appreciated.

**Comment by hgrsd** [comment]
@dkayiwa would you be the right person to ask for a code review so that we could move this forward? If not, is there anyone else that you could ask for a review?

**Comment by hgrsd** [comment]
@dkayiwa would you be the right person to ask for a code review so that we could move this forward? If not, is there anyone else that we could ask for a review? I'm keen not to let this get stale if possible.

**Comment by hgrsd** [comment]
@dkayiwa would you be the right person to ask for a code review so that we could move this forward? If not, is there anyone else that we could ask for a review? I'm keen not to let this get stale, if possible.

File: api/src/main/java/org/openmrs/api/DiagnosisService.java [hunk 3]
```diff
@@ -67,12 +68,40 @@
...
+	/**
+	 * Gets diagnoses for an Encounter. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param encounter the encounter for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
+	 * @param confirmedOnly whether to return only confirmed diagnoses
+	 * @return the list of (primary, confirmed) diagnoses for the given encounter
+	 */
+	@Authorized({ PrivilegeConstants.GET_DIAGNOSES })
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/DiagnosisService.java:80]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NewPublicMethodsandClasses

File: api/src/main/java/org/openmrs/api/DiagnosisService.java [hunk 4]
```diff
@@ -67,12 +68,40 @@
...
+	/**
+	 * Gets diagnoses for a Visit. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param visit the visit for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
+	 * @param confirmedOnly whether to return only confirmed diagnoses
+	 * @return the list of (primary, confirmed) diagnoses for the given encounter
+	 */
+	@Authorized({ PrivilegeConstants.GET_DIAGNOSES })
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/DiagnosisService.java:92]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NewPublicMethodsandClasses

File: api/src/main/java/org/openmrs/api/DiagnosisService.java [hunk 5]
```diff
@@ -67,12 +68,40 @@
...
+	 * Gets diagnoses for an Encounter. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param encounter the encounter for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
+	 * @param confirmedOnly whether to return only confirmed diagnoses
+	 * @return the list of (primary, confirmed) diagnoses for the given encounter
+	 */
+	@Authorized({ PrivilegeConstants.GET_DIAGNOSES })
+	List<Diagnosis> getDiagnosesForEncounter(Encounter encounter, boolean primaryOnly, boolean confirmedOnly);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/DiagnosisService.java:81]: Our convention has been `getDiagnosesByEncounter`. That is, `By` instead of `For`

File: api/src/main/java/org/openmrs/api/DiagnosisService.java [hunk 6]
```diff
@@ -67,12 +68,40 @@
...
+	 * Gets diagnoses for a Visit. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param visit the visit for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
+	 * @param confirmedOnly whether to return only confirmed diagnoses
+	 * @return the list of (primary, confirmed) diagnoses for the given encounter
+	 */
+	@Authorized({ PrivilegeConstants.GET_DIAGNOSES })
+	List<Diagnosis> getDiagnosesForVisit(Visit visit, boolean primaryOnly, boolean confirmedOnly);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/DiagnosisService.java:93]: `By` instead of `For`

File: api/src/main/java/org/openmrs/api/DiagnosisService.java [hunk 7]
```diff
@@ -67,12 +68,40 @@
...
+	List<Diagnosis> getDiagnosesForEncounter(Encounter encounter, boolean primaryOnly, boolean confirmedOnly);
+	
+	/**
+	 * Gets diagnoses for a Visit. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param visit the visit for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
+	 * @param confirmedOnly whether to return only confirmed diagnoses
+	 * @return the list of (primary, confirmed) diagnoses for the given encounter
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/DiagnosisService.java:90]: for the given encounter or visit?

File: api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java [hunk 8]
```diff
@@ -56,22 +57,39 @@
...
 	 */
+	@Deprecated
 	List<Diagnosis> getDiagnoses(Encounter encounter);
+	
+	/**
+	 * Gets diagnoses for an Encounter. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param encounter the encounter for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:77]: Instead of duplicating this description, shouldn't we just use? `@see org.openmrs.api.DiagnosisService#getDiagnosesByEncounter(org.openmrs.Encounter, boolean, boolean)`

File: api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java [hunk 9]
```diff
@@ -78,19 +78,49 @@ public Diagnosis getDiagnosisByUuid(String uuid) {
 		return diagnosisDAO.getActiveDiagnoses(patient, fromDate);
 	}
 
+	/**
+	 * Gets diagnoses for an Encounter. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param encounter the encounter for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java:86]: Use the `@see` annotation instead of duplicating all this.

File: api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java [hunk 10]
```diff
@@ -78,19 +78,49 @@ public Diagnosis getDiagnosisByUuid(String uuid) {
...
+	public List<Diagnosis> getDiagnosesForEncounter(Encounter encounter, boolean primaryOnly, boolean confirmedOnly) {
+		return diagnosisDAO.getDiagnosesForEncounter(encounter, primaryOnly, confirmedOnly);
+	}
+	
+	/**
+	 * Gets diagnoses for a Visit. When specified, this method only returns
+	 * primary or confirmed diagnoses.
+	 *
+	 * @param visit the visit for which to fetch diagnoses
+	 * @param primaryOnly whether to return only primary diagnoses
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java:101]: Use the `@see` annotation instead of duplicating all this.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 11]
```diff
@@ -92,13 +93,13 @@ public void shouldGetDiagnosisById() {
 
 	@Test
 	public void shouldGetActiveDiagnosesWithFromDate() {
-		Calendar calendar = new GregorianCalendar(2014,1,1,13,24,56);
-		assertEquals(1, diagnosisDAO.getActiveDiagnoses(new Patient(2), calendar.getTime()).size()); 
+		Calendar calendar = new GregorianCalendar(2014, 1, 1, 13, 24, 56);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java:96]: I would do these formatting changes in a separate pull request.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 12]
```diff
@@ -92,13 +93,13 @@ public void shouldGetDiagnosisById() {
...
+		assertEquals(1, diagnosisDAO.getActiveDiagnoses(new Patient(2), calendar.getTime()).size());
 	}
 
 	@Test
 	public void shouldGetActiveDiagnosesWithDifferentFromDate() {
-		Calendar calendar = new GregorianCalendar(2012,1,3,13,32,36);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java:101]: I would do these formatting changes in a separate pull request.

File: api/src/test/java/org/openmrs/api/impl/DiagnosisServiceImplTest.java [hunk 13]
```diff
@@ -41,34 +43,38 @@
...
 
 	@BeforeEach
-	public void setUp (){
-		if(diagnosisService == null){
+	public void setUp() {
+		if (diagnosisService == null) {
 			diagnosisService = Context.getDiagnosisService();
 		}
-		if(patientService == null){
+		if (patientService == null) {
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/impl/DiagnosisServiceImplTest.java:58]: I would do these formatting changes in a separate pull request.

**Comment by hgrsd** [comment]
Hi @dkayiwa - thanks for your comments, I appreciate it. I've made the changes you suggested and have undone some of the formatting changes so that the changeset is easier to follow.

Let me know if you're happy with this. If so, I can squash the commits :)

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 14]
```diff
@@ -100,18 +104,65 @@ public Diagnosis saveDiagnosis(Diagnosis diagnosis) {
 	}
 
 	/**
-	 * Gets primary diagnoses for a given encounter
+	 * Gets diagnoses for an Encounter. When specified, this method only returns
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:107]: What does this mean? `When specified, this method only returns primary or confirmed diagnoses`?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 15]
```diff
@@ -100,18 +104,65 @@ public Diagnosis saveDiagnosis(Diagnosis diagnosis) {
...
+		}
+		if (confirmedOnly) {
+			query.setParameter("certainty", ConditionVerificationStatus.CONFIRMED);
+		}
+
+		return query.getResultList();
+	}
+
+	/**
+	 * Gets diagnoses for an Encounter. When specified, this method only returns
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:138]: Did you forget about the `@see` annotation?

Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Yep, I did indeed - apologies.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 14 above)
Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: I was trying to express that optionally this method can only return primary or confirmed diagnoses. It's not very clear though so I removed the phrase. I think the arguments and the rest of the docstring make it clear enough how the method is to be used.

File: api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java [hunk 16]
```diff
@@ -56,22 +57,27 @@
...
 	 */
 	void deleteDiagnosis(Diagnosis diagnosis) throws DAOException;
-	
+
 	/**
 	 * Gets all diagnoses for a given encounter
 	 *
-	 * @param encounter the specific encounter to get the diagnoses for.
-	 * @return list of diagnoses for an encounter
+	 * @deprecated since 2.5.0, use {@link #getDiagnosesByEncounter}
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:64]: We do not deprecate methods in the data access layer because we do not expect clients to access them directly. They should use the service layer, which then indirectly accesses the data access layer. So we just remove them.

Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:64]: Thank you @dkayiwa. I had initially removed the method from the DAO altogether, but was asked to re-introduce it and deprecate it: https://github.com/openmrs/openmrs-core/pull/3831#discussion_r681121820.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java:64]: Sorry for making you run back and forth. You had done right to remove them.

Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/DiagnosisDAO.java]: No problem, I've updated the PR :)

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 17]
```diff
@@ -86,32 +90,53 @@ public Diagnosis saveDiagnosis(Diagnosis diagnosis) {
...
+	 * @see org.openmrs.api.db.DiagnosisDAO#getDiagnosesByEncounter(Encounter, boolean, boolean)
 	 */
 	@Override
-	public List<Diagnosis> getDiagnoses(Encounter encounter){
-		Query query = sessionFactory.getCurrentSession().createQuery(
-			"from Diagnosis d where d.encounter.encounterId = :encounterId order by dateCreated desc");
-		query.setInteger("encounterId", encounter.getId());
-		return query.list();	
+	public List<Diagnosis> getDiagnosesByEncounter(Encounter encounter, boolean primaryOnly, boolean confirmedOnly) {
+		String queryString = "from Diagnosis d where d.encounter.id = :encounterId";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:97]: Could there be a reason why you used `id` instead of the original `encounterId`?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 18]
```diff
@@ -86,32 +90,53 @@ public Diagnosis saveDiagnosis(Diagnosis diagnosis) {
...
 	 */
 	@Override
-	public List<Diagnosis> getPrimaryDiagnoses(Encounter encounter) {
-		Query query = sessionFactory.getCurrentSession().createQuery(
-			"from Diagnosis d where d.encounter.encounterId = :encounterId and d.rank = :rankId order by dateCreated desc");
-		query.setInteger("encounterId", encounter.getId());
-		query.setInteger("rankId", PRIMARY_RANK);
-		return query.list();
+	public List<Diagnosis> getDiagnosesByVisit(Visit visit, boolean primaryOnly, boolean confirmedOnly) {
+		String queryString = "from Diagnosis d where d.encounter.visit.id = :visitId";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:122]: Same as above for visitId

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 17 above)
Comment (hgrsd) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java:97]: No particular reason. I saw it done this way in the implementation of a different module (linked on the JIRA ticket) and used it. I didn't realise that the entities used `encounterId` and `visitId` as the PK field names, and the test passed so I didn't notice a discrepancy.

Will update it :)

**Comment by hgrsd** [comment]
Thanks @dkayiwa and everyone for guiding me through my first contribution to the project :-)

**Comment by dkayiwa** [comment]
Thanks @hgrsd for the awesome contribution! 👍 
We are please sorry for the delays in reviewing this.
We plan to do better for your next pull request. 😊
